numpy/core/src/npymath/npy_math_internal.h.src

```diff
@@ -716,9 +710,6 @@ npy_divmod@c@(@type@ a, @type@ b, @type@ *modulus)
     @type@ div, mod, floordiv;
 
     /* force set invalid flag, doesnt raise by default on gcc < 8 */
```

Comment:
This line should probably go, too.

Reply:
This is my "problem" so to speak.  I am confused what this is about and what we tried to do in gh-16161.

My guess is that [this godbolt snippet explains the comment](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:2,endLineNumber:27,positionColumn:2,positionLineNumber:27,selectionStartColumn:2,selectionStartLineNumber:27,startColumn:2,startLineNumber:27),source:'//+Type+your+code+here,+or+load+an+example.%0A%23include+%3Cmath.h%3E%0A%23include+%3Cfenv.h%3E%0A%23include+%3Cstdio.h%3E%0A%0Aint%0Atest(double+num)%0A%7B%0A++++if+(num+%3E+0)+%7B%0A++++++++return+1%3B%0A++++%7D%0A++++return+0%3B%0A%7D%0A%0A%0A%0Aint%0Amain(void)%0A%7B%0A++++double+value+%3D+NAN%3B%0A++++feclearexcept(FE_ALL_EXCEPT)%3B%0A%0A++++int+res+%3D+test(value)%3B%0A++++int+errs+%3D+fetestexcept(FE_ALL_EXCEPT)%3B%0A++++printf(%22Should+not+be+0,+should+have+an+exception+set:+%25d%5Cn%22,+errs)%3B%0A++++return+res%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:g73,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1'),fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'x86-64+gcc+7.3+(Editor+%231,+Compiler+%231)+C%2B%2B',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:14,fontUsePx:'0',wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+7.3+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) if you switch to GCC 8, the correct answer is given (warning flag is set). 

Now... If `a is NaN` and `b < 0`, the `if (mod)` will evaluate `True` if mod is NaN. Especially, if both `a` and `b` are `NaN`, then we definitely take that branch and both `b < 0` and `mod < 0` evaluate to False **with** setting the warning (since they are both NaN).  And that was *wrong* on GCC 7, which did not set the warning (but led to the *better* result).

(Hopefully, that flow analysis is right... But something is off and my best guess is to fix most of these comparisons and replace them with `isgreater` and `isless`, as mentioned before and now also in gh-19398.  In general, I don't think we really ever want float comparisons to set warning flags really, so de-facto *all* of those in the math code are probably just wrong...)

I honestly don't understand why:
```
with np.errstate(invalid='raise'):
    assert_no_warnings(FloatingPointError, np.floor_divide, 1, np.nan)
```
succeeds!  If I just call `np.floor_divide(1, np.nan)`, I get an "invalid value" warning, and with `np.errstate(invalid="raise")` the `FloatingPointError`.

Reply:
@arubiales I am confused, did you maybe try these things out on a gcc version older than 8 (Which would be pretty old)?

Reply:
@nschloe Yes sorry I fogot to delete it.  

@seberg No my GCC version is 8.4.  I try the same, and with the changes I have no warning, but I'm with you, should raise the "invalid value" flag. After test with this change:
```
NPY_INPLACE @type@
npy_floor_divide@c@(@type@ a, @type@ b) {
    @type@ div, mod;
    if (NPY_UNLIKELY(!b)) {
        div = a / b;
        if (!a || npy_isnan(a)) {
            npy_set_floatstatus_invalid();
        } else {
            npy_set_floatstatus_divbyzero();
        }
    } else {
        div = npy_divmod@c@(a, b, &mod);
        /* This is the change, I just set the invalid float status in the else */
        npy_set_floatstatus_invalid();
    }
```

 Which mean I never enter in the `if` because `NPY_UNLIKELY(!b)` is **always** `False`

Reply:
Yeah, `!b` is false, but then it goes into divmod?  I have to check more, maybe I was also missing a `git clean`...

Can you do me a favor and delete the `assert_no_warnings` completely and make it `np.errstate(all="raise")`?  I still have to think about it more.  I guess this is a good change though, and we should have more fixes as followup.